File: rag-backend/app/schemas/writing.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.db.models_writing import MAX_MAIN_TEX_CONTENT_CHARS
from app.db.writing_projects_repository import (
    MAX_REFERENCES_PER_PROJECT,
    AddReferenceOutcome,
)
from app.schemas.documents import BibliographicMetadataFields

_MAX_TITLE_CHARS = 200
_MAX_DESCRIPTION_CHARS = 2000

#: Milestone 5 Section 8 — a genuinely blank starting point: no fake
#: authors/affiliations/references (Section 8's explicit "do not include
#: fake authors, affiliations, references, or content"). `{title}` is
#: filled in with the project's own title at creation time.
#:
#: Milestone 5.1 Part 3/20 additions: `inputenc`/`fontenc`/`lmodern` give
#: baseline accented-Latin Unicode support (the compiler service's own
#: chosen pdflatex + utf8 combination — see latex-compiler/Dockerfile's
#: comment on why full multi-script Unicode is explicitly out of scope).
#: `\\bibliographystyle`/`\\bibliography{references}` are included from
#: project creation onward so Part 20's requirement ("the default project
#: must compile successfully... without user editing anything") and Part
#: 21 (inserting a citation and compiling it must resolve, "no fake
#: bibliography required") both hold without this service ever silently
#: rewriting a user's saved main_tex_content to inject them later — an
#: EXISTING project created before this milestone will not have these
#: two lines, and a `\\cite{}` typed into one will show the honest
#: "add \\bibliography{references} to resolve citations" diagnostic
#: (Part 22) rather than a citation silently failing to resolve for an
#: invisible reason.
DEFAULT_MAIN_TEX_TEMPLATE = """\\documentclass{{article}}
\\usepackage[utf8]{{inputenc}}
\\usepackage[T1]{{fontenc}}
\\usepackage{{lmodern}}

\\title{{{title}}}

\\begin{{document}}

\\maketitle

\\section{{Introduction}}

\\bibliographystyle{{plain}}
\\bibliography{{references}}

\\end{{document}}
"""


class WritingProjectSummaryResponse(BaseModel):
    id: str
    title: str
    description: str | None = None
    reference_count: int
    #: Milestone 5.3 Part 26 — total file/folder count in this project's
    #: tree, including the root .tex file (never references.bib, which
    #: is never a real row — see WritingProjectFile's own docstring).
    file_count: int = 0
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class WritingProjectListResponse(BaseModel):
    projects: list[WritingProjectSummaryResponse]
    total: int


class WritingProjectResponse(BaseModel):
    """The full project, including its LaTeX source — GET/POST/PATCH
    /writing-projects/{id} only (never the list response, see
    WritingProjectsRepository.list_for_user's own docstring)."""

    id: str
    title: str
    description: str | None = None
    main_tex_content: str
    #: Milestone 5.3 Part 15 — the project's current root/main `.tex`
    #: file id. Always populated for any project created or migrated
    #: under this milestone (see WritingProject.root_file_id's own
    #: docstring for the one theoretical transient exception).
    root_file_id: str | None = None
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class CreateWritingProjectRequest(BaseModel):
    title: str = Field(min_length=1, max_length=_MAX_TITLE_CHARS)
    description: str | None = Field(default=None, max_length=_MAX_DESCRIPTION_CHARS)


class DuplicateWritingProjectRequest(BaseModel):
    """Milestone 5.3 Part 29 — `title` is optional; when omitted the
    duplicate is titled "{original title} (copy)"."""

    title: str | None = Field(default=None, min_length=1, max_length=_MAX_TITLE_CHARS)


class UpdateWritingProjectRequest(BaseModel):
    """Partial update — same "only touch a field actually present"
    convention as UpdateProjectRequest/UpdateDocumentMetadataRequest.
    `title` can never be blanked; `description` can be explicitly
    cleared; `main_tex_content` is what the editor's autosave PATCHes on
    every debounced save (Section 7)."""

    title: str | None = Field(default=None, min_length=1, max_length=_MAX_TITLE_CHARS)
    description: str | None = Field(default=None, max_length=_MAX_DESCRIPTION_CHARS)
    main_tex_content: str | None = Field(default=None, max_length=MAX_MAIN_TEX_CONTENT_CHARS)


class WritingProjectReferenceResponse(BibliographicMetadataFields):
    """One project reference — the document's own CURRENT canonical
    metadata (Section 3), never a copied/frozen snapshot. `cited` reflects
    whether this document's citation_key currently appears anywhere in
    the project's main_tex_content (Section 27), computed fresh on every
    request from app/core/latex_citations.parse_cite_keys — never stored,
    so it can never drift out of sync with the actual manuscript text."""

    document_id: str
    source_filename: str
    added_at: datetime
    cited: bool


class WritingProjectReferencesResponse(BaseModel):
    references: list[WritingProjectReferenceResponse]
    total: int
    #: Citation keys parsed out of main_tex_content that do NOT match any
    #: current reference's citation_key (Section 26) — surfaced here so
    #: the frontend never needs a second parse of the same source text.
    missing_citation_keys: list[str] = Field(default_factory=list)
    #: Milestone 5.1 Part 34/35 — sha256(main_tex_content + references.bib),
    #: computed fresh on every call (app/core/latex_source_hash.py). The
    #: frontend compares this against the `source_hash` returned by its
    #: last successful compile to show "Preview is current" vs "Source
    #: changed since last compile" — piggybacked on this endpoint's
    #: existing "always fresh, refetched after every save" behavior (see
    #: [id].tsx's saveStatus-triggered reload) rather than a new polling
    #: endpoint.
    source_hash: str = ""


class AddWritingProjectReferenceRequest(BaseModel):
    document_ids: list[str] = Field(min_length=1, max_length=MAX_REFERENCES_PER_PROJECT)


class AddWritingProjectReferenceResult(BaseModel):
    document_id: str
    outcome: AddReferenceOutcome


class AddWritingProjectReferencesResponse(BaseModel):
    results: list[AddWritingProjectReferenceResult]


class WritingProjectBibliographyResponse(BaseModel):
    """GET /writing-projects/{id}/bibliography — generated fresh, on
    demand, from the project's current references via the exact same
    Milestone 4.2 BibTeX engine every other BibTeX surface uses (Section
    12: "Do NOT create manually duplicated BibTeX rows in another
    database")."""

    bibtex: str
    reference_count: int


class CompileDiagnosticResponse(BaseModel):
    severity: str
    message: str
    line: int | None = None
    # Milestone 5.5 Part 14 — the erroring file's project-relative path
    # (e.g. "sections/introduction.tex"), passed through verbatim from
    # the compiler service's own Diagnostic.file — see that field's
    # docstring for why it's only ever populated from the reliable
    # -file-line-error log form, never guessed.
    file: str | None = None


class CompileWritingProjectResponse(BaseModel):
    """Milestone 5.1 Part 15 — POST /writing-projects/{id}/compile's
    response. Deliberately never includes container paths, environment
    variables, or a raw stack trace (Part 15) — `log_excerpt` is already
    sanitized by the compiler service (latex-compiler/app/log_sanitizer.py)
    before it ever reaches this backend."""

    status: Literal["success", "error", "timeout", "busy", "unavailable"]
    diagnostics: list[CompileDiagnosticResponse] = Field(default_factory=list)
    log_excerpt: str = ""
    duration_ms: float = 0.0
    page_count: int | None = None
    #: Present only when status == "success" — the frontend fetches the
    #: actual PDF bytes via a SEPARATE
    #: GET /writing-projects/{id}/compile/{compile_id}/pdf call (Part 16:
    #: never base64-embed a PDF in this JSON response).
    compile_id: str | None = None
    pdf_size_bytes: int | None = None
    #: Milestone 5.1 Part 34/35 — the exact source_hash this compile was
    #: run against (app/core/latex_source_hash.py) — lets the frontend
    #: detect staleness without a version-history subsystem.
    source_hash: str = ""

File: packages/education-assistant-client/src/types/writing.ts
import type { components } from './generated';

/**
 * Milestone 5 (Academic Writing & LaTeX Foundation) — a LaTeX writing
 * project: one manuscript source plus its associated library references.
 * Deliberately a separate concept from Project/ProjectSummary
 * (types/projects.ts), which groups a user's CHAT CONVERSATIONS — see the
 * backend's app/db/models_writing.py docstring for the full reasoning.
 */
export type WritingProjectSummary = components['schemas']['WritingProjectSummaryResponse'];
export type WritingProjectListResponse = components['schemas']['WritingProjectListResponse'];
/**
 * The full project, including its LaTeX source — returned by
 * create/get/update only, never by the list endpoint (see
 * WritingProjectSummary — listing many projects must not transfer every
 * manuscript's full source just to render a summary card).
 */
export type WritingProject = components['schemas']['WritingProjectResponse'];

export interface CreateWritingProjectRequest {
  title: string;
  description?: string | null;
}

/**
 * Partial update — omit a field entirely to leave it unchanged (matches
 * the backend's PATCH /writing-projects/{id}, which only touches a field
 * actually present in the JSON body). `title` can never be cleared to
 * null. `mainTexContent` is what the editor's autosave PATCHes on a
 * debounced interval, never per keystroke.
 */
export interface UpdateWritingProjectRequest {
  title?: string;
  description?: string | null;
  mainTexContent?: string;
}

/**
 * One project reference — the associated document's own CURRENT
 * canonical bibliographic metadata (never a copied/frozen snapshot — see
 * the backend's WritingProjectReferenceResponse docstring). `cited`
 * reflects whether this document's citation key currently appears
 * anywhere in the project's main_tex_content, computed fresh on every
 * request.
 */
export type WritingProjectReference = components['schemas']['WritingProjectReferenceResponse'];
export type WritingProjectReferencesResponse =
  components['schemas']['WritingProjectReferencesResponse'];

/** Every outcome POST /writing-projects/{id}/references can report for
 * one document_id — never a hard failure for any of these ordinary
 * cases. */
export type AddWritingProjectReferenceOutcome =
  components['schemas']['AddWritingProjectReferenceResult']['outcome'];
export type AddWritingProjectReferenceResult =
  components['schemas']['AddWritingProjectReferenceResult'];
export type AddWritingProjectReferencesResponse =
  components['schemas']['AddWritingProjectReferencesResponse'];

/**
 * GET /writing-projects/{id}/bibliography — generated fresh, on demand,
 * from the project's current references via the exact same Milestone 4.2
 * BibTeX engine every other BibTeX surface uses. Never a second,
 * independently-maintained bibliography store.
 */
export type WritingProjectBibliography = components['schemas']['WritingProjectBibliographyResponse'];

/**
 * Milestone 5.1 (Secure LaTeX Compilation Service) — the structured
 * result of POST /writing-projects/{id}/compile. `status` covers both
 * genuine compile outcomes ("success" | "error" | "timeout") and the
 * backend's own honest "couldn't even try" outcomes ("busy" — the
 * compiler's bounded queue was full; "unavailable" — the compiler
 * service itself couldn't be reached) — never a raw HTTP error for any
 * of these, so the UI can render one consistent inline result.
 */
export type CompileStatus = 'success' | 'error' | 'timeout' | 'busy' | 'unavailable';
export type CompileDiagnostic = components['schemas']['CompileDiagnosticResponse'];
export type CompileWritingProjectResponse =
  components['schemas']['CompileWritingProjectResponse'];

/**
 * Milestone 5.3 (LaTeX Project Workspace & File Management) — one real
 * (database-backed) file or folder in a Writing Project's tree.
 * `references.bib` is NEVER represented here (see GeneratedFileNode
 * below) — it has no `id` and no mutation actions apply to it.
 */
export type WritingProjectFileNode = components['schemas']['WritingProjectFileNodeResponse'];
export type WritingProjectFileKind = WritingProjectFileNode['kind'];

/**
 * The synthesized, read-only `references.bib` tree entry — regenerated
 * fresh from the project's current references on every
 * GET /writing-projects/{id}/files call, never a real file row (Part 3:
 * "references.bib remains canonical... never a second source of
 * truth").
 */
export type GeneratedFileNode = components['schemas']['GeneratedFileNodeResponse'];

export type WritingProjectFileTree = components['schemas']['WritingProjectFileTreeResponse'];

/** GET /writing-projects/{id}/files/{fileId} — text files include their
 * content inline (`contentText`); binary files' bytes are fetched
 * separately via fetchWritingProjectFileContentBlob. */
export type WritingProjectFileContent =
  components['schemas']['WritingProjectFileContentResponse'];

/**
 * Bibliography Source Detection — how a Writing Project ACTUALLY
 * manages its citations/references, detected server-side from real
 * project content (see rag-backend's app/core/reference_mode.py) rather
 * than assumed from EduM8's own `references.bib` merely existing (it
 * always does, as a virtual/generated file — see GeneratedFileNode
 * above). One of:
 *
 * - "edum8_library" — EduM8's own generated references.bib.
 * - "imported_bib" — a real imported `.bib` database, e.g. a Springer
 *   Nature template's own "sn-bibliography.bib".
 * - "template_tex" — a template `.tex` file (e.g. "Bibliography.tex")
 *   with a manual `\begin{thebibliography}`/`\bibitem{...}` block.
 * - "inline_template" — the same, but directly inside the root/source
 *   `.tex` rather than a separate included file.
 */
export type ReferenceMode = components['schemas']['ReferenceModeResponse']['mode'];
export type CitationKeySource = components['schemas']['ReferenceModeResponse']['citation_key_source'];
export type ReferenceKey = components['schemas']['ReferenceKeyResponse'];
export type Edum8SwitchProposal = components['schemas']['Edum8SwitchProposalResponse'];
export type WritingProjectReferenceMode = components['schemas']['ReferenceModeResponse'];

export interface SwitchWritingProjectToEdum8ReferencesRequest {
  filePath: string;
  find: string;
  replace: string;
}

export interface CreateWritingProjectFolderRequest {
  parentId?: string | null;
  name: string;
}

export interface CreateWritingProjectTextFileRequest {
  parentId?: string | null;
  name: string;
  contentText?: string;
}

/**
 * Milestone 5.3 Part 39 — the multi-file autosave PATCH body. One
 * request, one file's content — never a batch update (keeps the
 * "flush current file before switching" discipline simple and matches
 * the existing PATCH /writing-projects/{id} main_tex_content
 * convention exactly).
 */
export interface UpdateWritingProjectFileContentRequest {
  contentText: string;
}

export interface RenameWritingProjectFileRequest {
  name: string;
}

export interface MoveWritingProjectFileRequest {
  /** null moves the file/folder to the project root. */
  newParentId: string | null;
}

export type WritingProjectFileMutationResponse =
  components['schemas']['WritingProjectFileMutationResponse'];
